syz-verifier/verifier.go

```diff
@@ -31,24 +36,180 @@ type Verifier struct {
 	// - <workdir>/corpus.db: corpus with interesting programs
 	// - <workdir>/<OS-Arch>/instance-x: per VM instance temporary files
 	// grouped by OS/Arch
-	workdir       string
-	crashdir      string
-	resultsdir    string
-	target        *prog.Target
-	runnerBin     string
-	executorBin   string
-	choiceTable   *prog.ChoiceTable
-	rnd           *rand.Rand
-	progIdx       int
-	addr          string
-	srv           *RPCServer
-	calls         map[*prog.Syscall]bool
-	reasons       map[*prog.Syscall]string
-	reportReasons bool
-	stats         *Stats
-	statsWrite    io.Writer
-	newEnv        bool
-	reruns        int
+	workdir           string
+	crashdir          string
+	resultsdir        string
+	target            *prog.Target
+	runnerBin         string
+	executorBin       string
+	progGeneratorInit sync.WaitGroup
+	choiceTable       *prog.ChoiceTable
+	rnd               *rand.Rand
+	progIdx           int
+	addr              string
+	srv               *RPCServer
+	calls             map[*prog.Syscall]bool
+	reasons           map[*prog.Syscall]string
+	reportReasons     bool
+	stats             *Stats
+	statsWrite        io.Writer
+	newEnv            bool
+	reruns            int
+
+	// We use single queue for every kernel environment.
+	tasksMutex     sync.Mutex
+	onTaskAdded    *sync.Cond
+	kernelEnvTasks [][]*ExecTaskQueue
```

Comment:
Now that we have priority, we don't need a queue per Environment. Environment is part of priority.

Reply:
Lets assume we have 2 environments - DirtyEnv and CleanEnv.
DirtyEnv is the VM that already executed some program(s).
CleanEnv is a fresh-booted VM that could run only 1 program. After that it become DirtyEnv.

I assume that we'll have a lot of the requests from DirtyEnv and will often wait for the CleanEnv.
It means, we can't just pop the task with highest priority and send it to DirtyEnv.
That's why I used array of heaps.

To remove one dimension here, we can use some binary tree based priority queue, but not a heap based one.